tests/library/test_json_schema.py

```diff
@@ -0,0 +1,661 @@
+                "other_str",
+                "my_A"
+            ],
+            "title": "B",
+            "type": "object"
+            }
+        },
+        "properties": {
+            "my_B": {
+            "$ref": "#/$defs/B"
+            }
+        },
+        "required": [
+            "my_B"
+        ],
+        "title": "C",
+        "type": "object"
+        }"""
+
+        target_obj = dict(my_B=dict(other_str="some string", my_A=dict(name="my name")))
+
+        # First sanity check what we're setting up
+        schema_obj = json.loads(schema)
+        validate(instance=target_obj, schema=schema_obj)
+
+        # The actual check
+        _generate_and_check(target_obj, schema_obj)
+
+
+class TestAnyOf:
+    @pytest.mark.parametrize("target_obj", [123, True])
+    def test_anyOf_simple(self, target_obj):
+        schema = """{
+        "anyOf": [
+                    {
+                        "type": "integer"
+                    },
+                    {
+                        "type": "boolean"
+                    }
+                ]
+    }
+    """
+        # First sanity check what we're setting up
+        schema_obj = json.loads(schema)
+        validate(instance=target_obj, schema=schema_obj)
+
+        # The actual check
+        _generate_and_check(target_obj, schema_obj)
+
+    @pytest.mark.parametrize(
+        "target_obj",
+        [
+            dict(my_val=dict(my_int=1)),
+            dict(my_val=dict(my_str="Some long string or other")),
+        ],
+    )
+    def test_anyOf_objects(self, target_obj):
+        schema = """{
+    "$defs": {
+        "A": {
+        "properties": {
+            "my_str": {
+            "default": "me",
+            "title": "My Str",
+            "type": "string"
+            }
+        },
+        "title": "A",
+        "type": "object"
+        },
+        "B": {
+        "properties": {
+            "my_int": {
+            "default": 1,
+            "title": "My Int",
+            "type": "integer"
+            }
+        },
+        "title": "B",
+        "type": "object"
+        }
+    },
+    "properties": {
+        "my_val": {
+        "anyOf": [
+            {
+            "$ref": "#/$defs/A"
+            },
+            {
+            "$ref": "#/$defs/B"
+            }
+        ],
+        "title": "My Val"
+        }
+    },
+    "title": "C",
+    "type": "object"
+    }
+    """
+        # First sanity check what we're setting up
+        schema_obj = json.loads(schema)
+        validate(instance=target_obj, schema=schema_obj)
+
+        # The actual check
+        _generate_and_check(target_obj, schema_obj)
+
+
+class TestRecursiveStructures:
+    @pytest.mark.parametrize(
+        "target_obj",
+        [
+            dict(my_list=None),
+            dict(my_list=dict(my_str="a", next=None)),
+            dict(my_list=dict(my_str="a", next=dict(my_str="b", next=None))),
+        ],
+    )
+    def test_linked_list(self, target_obj):
+        schema = """
```

Comment:
@hudson-ai  did you say you had a linked list working with the library functions? This is blowing up recursively on me still, even with the various `lm` bits intervening.